src/server/pps/server/api_server.go

```diff
@@ -2492,37 +2492,79 @@ func (a *apiServer) listPipeline(pachClient *client.APIClient, request *pps.List
 		// ensure field names and enum values match with --raw output
 		enc = serde.NewJSONEncoder(&jsonBuffer, serde.WithOrigName(true))
 	}
-	return a.listPipelinePtr(pachClient, request.Pipeline, request.History,
-		func(name string, ptr *pps.EtcdPipelineInfo) error {
-			var pipelineInfo *pps.PipelineInfo
-			var err error
-			if request.AllowIncomplete {
-				pipelineInfo, err = ppsutil.GetPipelineInfoAllowIncomplete(pachClient, name, ptr)
+	filterPipeline := func(pipelineInfo *pps.PipelineInfo) bool {
+		if jqCode != nil {
+			jsonBuffer.Reset()
+			// convert pipelineInfo to a map[string]interface{} for use with gojq
+			enc.EncodeProto(pipelineInfo)
+			var pipelineInterface interface{}
+			json.Unmarshal(jsonBuffer.Bytes(), &pipelineInterface)
+			iter := jqCode.Run(pipelineInterface)
+			// treat either jq false-y value as rejection
+			if v, _ := iter.Next(); v == false || v == nil {
+				return false
+			}
+		}
+		return true
+	}
+	// the mess below is so we can lookup the PFS info for each pipeline concurrently.
+	type etcdInfo struct {
+		name string
+		ptr  *pps.EtcdPipelineInfo
+	}
+	eg, ctx := errgroup.WithContext(pachClient.Ctx())
+	etcdInfos := make(chan etcdInfo)
+	// stream these out of etcd
+	eg.Go(func() error {
+		defer close(etcdInfos)
+		return a.listPipelinePtr(pachClient, request.Pipeline, request.History, func(name string, ptr *pps.EtcdPipelineInfo) error {
+			// copy
+			data, _ := ptr.Marshal()
+			ptr2 := &pps.EtcdPipelineInfo{}
+			if err := ptr2.Unmarshal(data); err != nil {
+				panic(err)
+			}
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case etcdInfos <- etcdInfo{name: name, ptr: ptr2}:
+				return nil
+			}
+		})
+	})
+	// spin up goroutines to get the PFS info, and then use the mutex to call f all synchronized like.
+	var mu sync.Mutex
+	for i := 0; i < 20; i++ {
+		eg.Go(func() error {
+			for info := range etcdInfos {
+				pinfo, err := a.resolvePipelineInfo(pachClient, request.AllowIncomplete, info.name, info.ptr)
 				if err != nil {
 					return err
 				}
-			} else {
-				pipelineInfo, err = ppsutil.GetPipelineInfo(pachClient, name, ptr)
-				if err != nil {
+				if err := func() error {
+					mu.Lock()
+					defer mu.Unlock()
+					// the filtering shares that buffer thing, and it's CPU bound so why not do it with the lock
+					if !filterPipeline(pinfo) {
+						return nil
+					}
+					return f(pinfo)
```

Comment:
The only problem I see is that we may call `f` after it has already errored.  Perhaps we should have a bool to bail out after acquiring `mu` if an error has occurred (or the `eg` has been canceled)?

Reply:
Is there a guarantee that f won't be called after it errors?  f won't be called after the function returns.  But I think the caller already doesn't know how many times f will be called.

To get the behavior you're describing, I guess we add another channel and then check the context for cancellation?